File: references/EMS/sweep_governor_scale.m
%% SWEEP_GOVERNOR_SCALE
%  How does the governor penalty depend on the current-threshold scale S_I?
%
%  S_I multiplies ONLY the Class-2 constants -- the six absolute-current
%  thresholds (min-load gate, minority floor, open-loop hysteresis, dark/live,
%  cut ceiling). The dimensionless droop band [0.15, 0.85] and the Class-3
%  rates (slew ceiling, dwell) are untouched by construction, so this sweep
%  isolates one mechanism instead of scaling the whole governor.
%
%  Why it matters: S_I = 127.8 rests on an inferred 2.0 A bench nominal
%  current, the least defensible number in the study. At S_I = 127.8 the
%  closed-loop entry threshold is 55 kW, which UDDS rarely reaches, so the
%  loop is locked out and the split is set by contactors. At S_I = 1 the
%  thresholds are milliwatt-scale, the loop is always closed, and only the
%  band and the slew limiter remain active.
%
%  If the penalty tracks S_I, the earlier "the governor is expensive when its
%  own entry threshold locks it out" reading is confirmed and the whole result
%  hinges on that one unverified constant.
%
%  Requires: SDP_EnergyManagement_Governor3.m, the UDDS cycle, TPM.

clear; close all; clc;

%% ------------------------------------------------------------------
%  1. Cycle and TPM
%  ------------------------------------------------------------------
data2 = load('simulink_pdem_output_UDDS.mat', 'out');
P_dem = data2.out.simout.Data;
Time  = data2.out.simout.Time;
wholeSeconds = 0:1:1369;
P_dem1 = interp1(Time, P_dem, wholeSeconds, 'linear');
P_dem1 = P_dem1(:).';

if ~exist('TPM', 'var')
    S = load('TPM.mat');
    f = fieldnames(S);
    TPM = S.(f{1});
end

SOC_initial = 0.6;
Em = 720;  Q = 100;

fprintf('UDDS: P_dem range [%.1f, %.1f] kW\n', min(P_dem1)/1e3, max(P_dem1)/1e3);

%% ------------------------------------------------------------------
%  2. Fixed configuration -- everything except S_I and alpha
%  ------------------------------------------------------------------
cfg = struct();
cfg.n_subticks        = 880;
cfg.P_fc_ramp_W_per_s = 50000;
cfg.h2_model          = 'convex';
cfg.h2_a0             = 0.05;
cfg.h2_P_peak_W       = 35000;
cfg.h2_eta_peak       = 0.50;
cfg.h2_start_cost_g   = 0.5;
cfg.h2_s_eq_fixed     = 1/(0.50*120000);
cfg.verbose           = false;

sweep_SI    = [1 10 25 50 100 127.8];
sweep_alpha = [200 500];

nS = numel(sweep_SI);
nA = numel(sweep_alpha);

res = struct();

%% ------------------------------------------------------------------
%  3. Sweep
%     The SDP policy does not depend on S_I, so it is computed once per
%     alpha and reused for every governor scale via cfg.policy_cache.
%     The baseline likewise does not depend on S_I -- one run per alpha.
%  ------------------------------------------------------------------
for ia = 1:nA

    a = sweep_alpha(ia);
    fprintf('\n===== alpha = %d =====\n', a);

    % ---- baseline: governor off, so S_I is irrelevant ---------------
    c_b = cfg;  c_b.alpha = a;  c_b.governor_enabled = false;
    fprintf('  baseline ... ');  t0 = tic;
    [~, ~, SOC_b, gb] = SDP_EnergyManagement_Governor3(P_dem1, SOC_initial, TPM, c_b);
    fprintf('%.1f s\n', toc(t0));

    cache = gb.policy_cache;              % reuse for every governed run
    s_eq  = gb.h2.s_eq;
    Eb_b  = -(SOC_b(end) - SOC_initial) * Em * 3600 * Q;

    res(ia).alpha       = a;
    res(ia).MH2_eq_base = gb.summary.M_H2_total + Eb_b*s_eq;
    res(ia).SOCrms_base = gb.summary.SOC_rms_dev;
    res(ia).starts_base = gb.summary.n_starts;

    % ---- governed, one run per S_I ----------------------------------
    for is = 1:nS
        SI = sweep_SI(is);
        c_g = cfg;
        c_g.alpha             = a;
        c_g.governor_enabled  = true;
        c_g.S_I               = SI;        % <-- the swept quantity
        c_g.policy_cache      = cache;     % skip value iteration

        fprintf('  S_I = %6.1f ... ', SI);  t0 = tic;
        [~, ~, SOC_g, gg] = SDP_EnergyManagement_Governor3(P_dem1, SOC_initial, TPM, c_g);
        fprintf('%.1f s\n', toc(t0));

        Eb_g = -(SOC_g(end) - SOC_initial) * Em * 3600 * Q;

        res(ia).SI(is)          = SI;
        res(ia).MH2_eq(is)      = gg.summary.M_H2_total + Eb_g*s_eq;
        res(ia).penalty(is)     = 100*(res(ia).MH2_eq(is) - res(ia).MH2_eq_base) ...
                                  / res(ia).MH2_eq_base;
        res(ia).SOCrms(is)      = gg.summary.SOC_rms_dev;
        res(ia).SOCratio(is)    = gg.summary.SOC_rms_dev / res(ia).SOCrms_base;
        res(ia).closed(is)      = gg.summary.frac_closed_loop;
        res(ia).latched(is)     = gg.summary.frac_latched;
        res(ia).events(is)      = gg.summary.n_latch_events;
        res(ia).mean_dPfc(is)   = gg.summary.mean_abs_dP_fc;
        res(ia).starts(is)      = gg.summary.n_starts;
        res(ia).sat(is)         = gg.summary.frac_saturated;
        res(ia).entry_kW(is)    = 2*gg.C.MINORITY_I_MIN_A*720/1e3;
    end
end

%% ------------------------------------------------------------------
%  4. Tables
%  ------------------------------------------------------------------
for ia = 1:nA
    r = res(ia);
    fprintf('\n========== SDP, UDDS, alpha = %d: governor scale sweep ==========\n', r.alpha);
    fprintf('baseline M_H2,eq = %.2f g,  SOC RMS = %.3e,  starts = %d\n\n', ...
            r.MH2_eq_base, r.SOCrms_base, r.starts_base);
    fprintf('%8s %10s %10s %9s %8s %8s %8s %8s %9s\n', ...
            'S_I','entry kW','MH2eq [g]','penalty','SOC x','closed','latched','events','md|dPfc|');
    for is = 1:nS
        fprintf('%8.1f %10.1f %10.2f %8.2f%% %8.2f %7.1f%% %7.1f%% %8d %9.0f\n', ...
                r.SI(is), r.entry_kW(is), r.MH2_eq(is), r.penalty(is), ...
                r.SOCratio(is), 100*r.closed(is), 100*r.latched(is), ...
                r.events(is), r.mean_dPfc(is));
    end
    fprintf('================================================================\n');
end

%% ------------------------------------------------------------------
%  5. Plots
%  ------------------------------------------------------------------
figure(1); clf
subplot(1,3,1)
for ia = 1:nA
    semilogx(res(ia).SI, res(ia).penalty, 'o-', 'LineWidth', 1.4); hold on
end
yline(0,'k:'); xlabel('S_I'); ylabel('governor penalty (%)'); grid on
legend(arrayfun(@(r) sprintf('\\alpha = %d', r.alpha), res, 'uni', 0), 'Location','best')
title('Fuel penalty vs current scale')

subplot(1,3,2)
for ia = 1:nA
    semilogx(res(ia).SI, 100*res(ia).closed, 'o-', 'LineWidth', 1.4); hold on
    semilogx(res(ia).SI, 100*res(ia).latched, 's--', 'LineWidth', 1.2);
end
xlabel('S_I'); ylabel('% of cycle'); grid on
legend('closed loop','latched','Location','best')
title('Governor operating regime')

subplot(1,3,3)
for ia = 1:nA
    semilogx(res(ia).SI, res(ia).SOCratio, 'o-', 'LineWidth', 1.4); hold on
end
yline(1,'k:'); xlabel('S_I'); ylabel('SOC RMS ratio (gov / base)'); grid on
title('SOC regulation')

figure(2); clf
for ia = 1:nA
    plot(100*res(ia).closed, res(ia).penalty, 'o', 'MarkerSize', 9, ...
         'MarkerFaceColor','auto'); hold on
    for is = 1:nS
        text(100*res(ia).closed(is), res(ia).penalty(is), ...
             sprintf('  %.0f', res(ia).SI(is)), 'FontSize', 8);
    end
end
yline(0,'k:'); grid on
xlabel('closed-loop fraction (%)'); ylabel('governor penalty (%)')
title('Penalty vs lockout (labels are S_I)')
legend(arrayfun(@(r) sprintf('\\alpha = %d', r.alpha), res, 'uni', 0), 'Location','best')

File: tools/fullscale_governor/sweep_governor_scale.py
#!/usr/bin/env python3
"""sweep_governor_scale.py - port of references/EMS/sweep_governor_scale.m.

Sweeps the governor current scale S_I (Class-2 constants only) for the SDP EMS
on the UDDS cycle at alpha in {200, 500} and prints the student's table:
S_I, closed-loop entry [kW], M_H2,eq governed [g], penalty vs the ungoverned
baseline, SoC RMS degradation ratio, closed-loop fraction.

Usage
    python sweep_governor_scale.py --pdem <file> [--tpm <mat>] [--alpha 200 500]
                                   [--si 1 10 25 50 100 127.8] [--json out.json]

--pdem accepts a .npy (1-D W at 1 Hz), a .mat with variable P_dem1, or the
student's simulink_pdem_output_UDDS.mat (decoded through tools/tpm_generator's
MCOS reader; resampled to whole seconds 0..1369 like the MATLAB).
"""
from __future__ import annotations

import argparse
import json
import os
import sys
import time

import numpy as np

HERE = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.abspath(os.path.join(HERE, "..", ".."))
sys.path.insert(0, HERE)
sys.path.insert(0, os.path.join(REPO, "tools"))

from sdp_governor3 import sdp_governor3, mh2_eq, EM_V, Q_AH  # noqa: E402

DEFAULT_TPM = os.path.join(REPO, "references", "EMS", "TPM_fullsize.mat")
SOC_INITIAL = 0.6
FIXED_CFG = dict(
    n_subticks=880, P_fc_ramp_W_per_s=50000.0, h2_model="convex", h2_a0=0.05,
    h2_P_peak_W=35000.0, h2_eta_peak=0.50, h2_start_cost_g=0.5,
    h2_s_eq_fixed=1.0 / (0.50 * 120000.0), verbose=False,
)


def load_pdem(path: str) -> np.ndarray:
    if path.endswith(".npy"):
        return np.load(path).astype(float).ravel()
    from scipy.io import loadmat
    m = loadmat(path)
    if "P_dem1" in m:
        return np.asarray(m["P_dem1"], dtype=float).ravel()
    # Simulink output: decode the opaque timeseries like tools/tpm_generator.py
    import tpm_generator as T
    meta = T._read_object_metadata(path)
    pairs = list(T._extract_timeseries_pairs(meta))
    P_pairs = [p for p in pairs if p[2].max() > 3e4 and p[2].min() < -1e4]
    if not P_pairs:
        raise ValueError("no P_dem-like timeseries in %s" % path)
    t, d = P_pairs[-1][1], P_pairs[-1][2]
    whole = np.arange(0, 1370.0)
    return np.interp(whole, t, d)          # interp1 linear onto 0:1:1369


def load_tpm(path: str) -> np.ndarray:
    from scipy.io import loadmat
    m = loadmat(path)
    key = [k for k in m if not k.startswith("__")][0]
    return np.asarray(m[key], dtype=float)


def run_sweep(P_dem, TPM, alphas, sweep_SI, log=print):
    res = []
    for a in alphas:
        log("\n===== alpha = %g =====" % a)
        cfg_b = dict(FIXED_CFG, alpha=a, governor_enabled=False)
        t0 = time.time()
        rb = sdp_governor3(P_dem, SOC_INITIAL, TPM, cfg_b)
        log("  baseline ... %.1f s (value iteration %d sweeps)" % (time.time() - t0, rb.vi_sweeps))
        row = dict(alpha=a, MH2_base=rb.summary["M_H2_total"], SOC_end_base=float(rb.SOC[-1]),
                   MH2_eq_base=mh2_eq(rb, SOC_INITIAL), SOCrms_base=rb.summary["SOC_rms_dev"],
                   starts_base=rb.n_starts, vi_sweeps=rb.vi_sweeps,
                   SI=[], entry_kW=[], MH2=[], SOC_end=[], MH2_eq=[], penalty=[], SOCrms=[], SOCratio=[],
                   closed=[], latched=[], events=[], mean_dPfc=[], starts=[], sat=[])
        for SI in sweep_SI:
            cfg_g = dict(FIXED_CFG, alpha=a, governor_enabled=True, S_I=SI, policy_cache=rb.policy_cache)
            t0 = time.time()
            rg = sdp_governor3(P_dem, SOC_INITIAL, TPM, cfg_g)
            log("  S_I = %6.1f ... %.1f s" % (SI, time.time() - t0))
            eq = mh2_eq(rg, SOC_INITIAL)
            row["SI"].append(SI)
            row["entry_kW"].append(2 * rg.C.MINORITY_I_MIN_A * EM_V / 1e3)
            row["MH2"].append(rg.summary["M_H2_total"])
            row["SOC_end"].append(float(rg.SOC[-1]))
            row["MH2_eq"].append(eq)
            row["penalty"].append(100.0 * (eq - row["MH2_eq_base"]) / row["MH2_eq_base"])
            row["SOCrms"].append(rg.summary["SOC_rms_dev"])
            row["SOCratio"].append(rg.summary["SOC_rms_dev"] / row["SOCrms_base"])
            row["closed"].append(rg.summary["frac_closed_loop"])
            row["latched"].append(rg.summary["frac_latched"])
            row["events"].append(rg.n_latch_events)
            row["mean_dPfc"].append(rg.summary["mean_abs_dP_fc"])
            row["starts"].append(rg.n_starts)
            row["sat"].append(rg.summary["frac_saturated"])
        res.append(row)
    return res


def print_tables(res, log=print):
    for r in res:
        log("\n========== SDP, UDDS, alpha = %g: governor scale sweep ==========" % r["alpha"])
        log("baseline M_H2,eq = %.2f g,  SOC RMS = %.3e,  starts = %d\n" % (r["MH2_eq_base"], r["SOCrms_base"], r["starts_base"]))
        log("%8s %10s %10s %9s %8s %8s %8s %8s %9s" % ("S_I", "entry kW", "MH2eq [g]", "penalty", "SOC x", "closed", "latched", "events", "md|dPfc|"))
        for i in range(len(r["SI"])):
            log("%8.1f %10.1f %10.2f %8.2f%% %8.2f %7.1f%% %7.1f%% %8d %9.0f" % (
                r["SI"][i], r["entry_kW"][i], r["MH2_eq"][i], r["penalty"][i], r["SOCratio"][i],
                100 * r["closed"][i], 100 * r["latched"][i], r["events"][i], r["mean_dPfc"][i]))
        log("=" * 64)


def main(argv=None):
    ap = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    ap.add_argument("--pdem", required=True)
    ap.add_argument("--tpm", default=DEFAULT_TPM)
    ap.add_argument("--alpha", type=float, nargs="+", default=[200.0, 500.0])
    ap.add_argument("--si", type=float, nargs="+", default=[1, 10, 25, 50, 100, 127.8])
    ap.add_argument("--json", default=None)
    a = ap.parse_args(argv)
    P_dem = load_pdem(a.pdem)
    TPM = load_tpm(a.tpm)
    print("UDDS: P_dem range [%.1f, %.1f] kW, N = %d" % (P_dem.min() / 1e3, P_dem.max() / 1e3, len(P_dem)))
    res = run_sweep(P_dem, TPM, a.alpha, a.si)
    print_tables(res)
    if a.json:
        with open(a.json, "w", encoding="utf-8") as fh:
            json.dump(res, fh, indent=1)
        print("wrote", a.json)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
